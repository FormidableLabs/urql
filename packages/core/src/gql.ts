/* eslint-disable prefer-rest-params */
import { TypedDocumentNode } from '@graphql-typed-document-node/core';

import {
  DocumentNode,
  DefinitionNode,
  FragmentDefinitionNode,
  Kind,
} from 'graphql';

import { keyDocument, stringifyDocument } from './utils';

const applyDefinitions = (
  fragmentNames: Map<string, string>,
  target: DefinitionNode[],
  source: Array<DefinitionNode> | ReadonlyArray<DefinitionNode>
) => {
  for (let i = 0; i < source.length; i++) {
    const definition = source[i];
    if (source[i].kind === Kind.FRAGMENT_DEFINITION) {
      const name = (source[i] as FragmentDefinitionNode).name.value;
      const value = stringifyDocument(definition);
      // Fragments will be deduplicated according to this Map
      const prevValue = fragmentNames.get(name);
      if (prevValue === undefined) {
        fragmentNames.set(name, value);
        target.push(source[i]);
      } else if (process.env.NODE_ENV !== 'production' && prevValue !== value) {
        // Fragments with the same names is expected to have the same contents
        console.warn(
          '[WARNING: Duplicate Fragment] A fragment with name `' +
            name +
            '` already exists in this document.\n' +
            'While fragment names may not be unique across your source, each name must be unique per document.'
        );
      }
    } else {
      target.push(source[i]);
    }
  }
};

function gql<Data = any, Variables = object>(
  strings: TemplateStringsArray,
  ...interpolations: Array<TypedDocumentNode | DocumentNode | string>
): TypedDocumentNode<Data, Variables>;

function gql<Data = any, Variables = object>(
  string: string
): TypedDocumentNode<Data, Variables>;

function gql(/* arguments */) {
  const fragmentNames = new Map<string, string>();
  const definitions: DefinitionNode[] = [];
  const interpolations: DocumentNode[] = [];

  // Apply the entire tagged template body's definitions
  let body: string = Array.isArray(arguments[0])
    ? arguments[0][0]
    : arguments[0] || '';
  for (let i = 1; i < arguments.length; i++) {
    const value = arguments[i];
    if (value && value.kind === Kind.DOCUMENT) {
      interpolations.push(value);
    } else {
      body += value;
    }

    body += arguments[0][i];
  }

  // Apply the tag's body definitions
  applyDefinitions(fragmentNames, definitions, keyDocument(body).definitions);

  // Copy over each interpolated document's definitions
  for (let i = 0; i < interpolations.length; i++)
    applyDefinitions(fragmentNames, definitions, interpolations[i].definitions);

  return keyDocument({
    kind: Kind.DOCUMENT,
    definitions,
  });
}

export { gql };
