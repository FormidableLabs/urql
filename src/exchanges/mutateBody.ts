import { pipe, tap, map } from 'wonka';
import { Exchange, Operation, OperationResult } from '../types';
import {
  DocumentNode,
  print,
  ASTNode,
  buildClientSchema,
  visitWithTypeInfo,
  TypeInfo,
  FragmentDefinitionNode,
} from 'graphql';
import { parse, visit, NonNullTypeNode } from 'graphql';
import { type } from 'os';
import { object } from 'prop-types';
import gql from 'graphql-tag';

type TypeFragmentMap<T extends string = string> = Record<string, string[]> & {
  _fragments?: FragmentDefinitionNode[];
};

interface ExchangeArgs {
  schema: any;
}

/** An exchange for auto-populating mutations with a required response body. */
export const mutateBodyExchange = ({ schema }: ExchangeArgs): Exchange => ({
  forward,
}) => {
  let typeFragments: TypeFragmentMap = { _fragments: [] };

  const handleIncomingMutation = (op: Operation) => {
    if (op.operationName !== 'mutation') {
      return op;
    }

    return;
  };

  const handleIncomingQuery = (op: Operation) => {
    if (op.operationName !== 'query') {
      return;
    }

    typeFragments = makeFragmentsFromQuery({
      schema,
      query: op.query,
      fragmentMap: typeFragments,
    });
  };

  return ops$ => {
    return pipe(
      ops$,
      tap(handleIncomingQuery),
      forward
    );
  };
};

interface MakeFragmentsFromQueryArg {
  schema: any;
  query: DocumentNode;
  fragmentMap: TypeFragmentMap;
}

/** Creates fragments object from query */
export const makeFragmentsFromQuery = ({
  schema,
  query,
  fragmentMap,
}: MakeFragmentsFromQueryArg) => {
  let f = fragmentMap;
  const typeInfo = new TypeInfo(buildClientSchema(schema));

  visit(
    query,
    visitWithTypeInfo(typeInfo, {
      Field: (node, key, parent, path) => {
        if (!node.selectionSet) {
          return undefined;
        }

        // @ts-ignore
        const t = typeInfo.getType().ofType;

        if (!t) {
          return undefined;
        }

        // @ts-ignore
        f = {
          ...f,
          [t]: [...(f[t] || []), node.selectionSet],
        };
      },
      FragmentDefinition: node => {
        // @ts-ignore
        f = {
          ...f,
          _fragments: [...(f._fragments || []), node],
        };
      },
    })
  );

  return f;
};

interface AddFragmentsToQuery {
  schema: any;
  query: DocumentNode;
  fragmentMap: TypeFragmentMap;
}

export const addFragmentsToQuery = ({
  schema,
  query,
  fragmentMap,
}: AddFragmentsToQuery) => {
  const typeInfo = new TypeInfo(buildClientSchema(schema));

  const x = visit(
    query,
    visitWithTypeInfo(typeInfo, {
      Field: {
        enter: node => {
          const directive =
            node.directives &&
            node.directives.find(d => d.name.value === 'populate');

          if (!directive) {
            return;
          }

          // @ts-ignore
          const type = typeInfo.getType().ofType;

          return {
            ...node,
            directives: (node.directives as any).filter(
              d => d.name.value !== 'populate'
            ),
            selectionSet: {
              kind: 'SelectionSet',
              selections: fragmentMap[type].map(selectionSet => ({
                kind: 'InlineFragment',
                typeCondition: {
                  kind: 'NamedType',
                  // @ts-ignore
                  name: { kind: 'Name', value: typeInfo.getType().ofType },
                },
                // @ts-ignore
                selectionSet,
              })),
            },
          };
        },
      },
      Document: {
        leave: node => {
          return {
            ...node,
            definitions: [
              ...node.definitions,
              ...(fragmentMap._fragments || []),
            ],
          };
        },
      },
    })
  );

  console.log(print(x));
};
