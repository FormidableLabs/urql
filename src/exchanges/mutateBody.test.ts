import { makeFragmentsFromQuery, addFragmentsToQuery } from './mutateBody';
import {
  introspectionFromSchema,
  buildSchema,
  print,
  visit,
  FragmentDefinitionNode,
  SelectionSetNode,
} from 'graphql';
import gql from 'graphql-tag';

const schemaDef = `
  type User {
    id: ID!
    name: String!
    age: Int!
  }

  type Todo {
    id: ID!
    text: String!
    creator: User!
  }

  type Query {
    todos: [Todo]
  }

  type Mutation {
    addTodo: [Todo]
  }
`;

const query = gql`
  query {
    todos {
      id
      name
      creator {
        id
      }
    }
  }

  fragment TodoFragment on Todo {
    id
  }
`;

const schema = introspectionFromSchema(buildSchema(schemaDef));

describe('makeFragmentsFromQuery', () => {
  const arg = {
    schema,
    query,
    fragmentMap: {
      _fragments: [
        gql`
          fragment MyFragment on User {
            id
            name
          }
        `.definitions[0],
      ],
    },
  };

  describe('new fragments', () => {
    it('are created', () => {
      const r = makeFragmentsFromQuery(arg);
      expect(r.Todo.length).toBe(1);
      expect(r).toMatchSnapshot();
    });
  });

  describe('user declared fragments', () => {
    it('are extracted', () => {
      const r = makeFragmentsFromQuery(arg);
      expect(
        (r._fragments as any[]).filter(f => f.name.value === 'TodoFragment')
      ).toHaveLength(1);
    });
  });
});

describe('addFragmentsToQuery', () => {
  const mutation = gql`
    mutation MyMutation {
      addTodo @populate
    }
  `;

  const fragmentMap = {
    Todo: [query.definitions[0].selectionSet.selections[0].selectionSet],
    _fragments: [
      gql`
        fragment MyFragment on User {
          id
          name
        }
      `,
    ],
  } as any;

  const arg = {
    schema,
    query: mutation,
    fragmentMap,
  };

  it('returns result matching snapshot', () => {
    const q = addFragmentsToQuery(arg);
    expect(print(q)).toMatchInlineSnapshot(`
      "mutation MyMutation {
        addTodo {
          ... on Todo {
            id
            name
            creator {
              id
            }
          }
        }
      }

      fragment MyFragment on User {
        id
        name
      }

      "
    `);
  });

  describe('user provided fragments', () => {
    it('are added to query', () => {
      const r = addFragmentsToQuery(arg);

      let fragmentDefs: FragmentDefinitionNode[] = [];
      visit(r, {
        FragmentDefinition: node => {
          fragmentDefs = [...fragmentDefs, node];
        },
      });

      expect(
        fragmentDefs.filter(d => d.name.value === 'MyFragment')
      ).toHaveLength(1);
    });
  });

  describe('new selection sets', () => {
    it('are added to query', () => {
      const r = addFragmentsToQuery(arg);

      let selelctionSets: SelectionSetNode[] = [];
      visit(r, {
        SelectionSet: node => {
          selelctionSets = [...selelctionSets, node];
        },
      });

      expect(
        selelctionSets.filter(s => s === fragmentMap.Todo[0])
      ).toHaveLength(1);
    });
  });
});
