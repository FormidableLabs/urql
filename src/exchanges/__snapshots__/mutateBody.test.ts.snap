// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeFragmentsFromQuery new fragments are created 1`] = `
Object {
  "Todo": Array [
    Object {
      "kind": "SelectionSet",
      "selections": Array [
        Object {
          "alias": undefined,
          "arguments": Array [],
          "directives": Array [],
          "kind": "Field",
          "name": Object {
            "kind": "Name",
            "value": "id",
          },
          "selectionSet": undefined,
        },
        Object {
          "alias": undefined,
          "arguments": Array [],
          "directives": Array [],
          "kind": "Field",
          "name": Object {
            "kind": "Name",
            "value": "name",
          },
          "selectionSet": undefined,
        },
        Object {
          "alias": undefined,
          "arguments": Array [],
          "directives": Array [],
          "kind": "Field",
          "name": Object {
            "kind": "Name",
            "value": "creator",
          },
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "id",
                },
                "selectionSet": undefined,
              },
            ],
          },
        },
      ],
    },
  ],
  "User": Array [
    Object {
      "kind": "SelectionSet",
      "selections": Array [
        Object {
          "alias": undefined,
          "arguments": Array [],
          "directives": Array [],
          "kind": "Field",
          "name": Object {
            "kind": "Name",
            "value": "id",
          },
          "selectionSet": undefined,
        },
      ],
    },
  ],
  "_fragments": Array [
    Object {
      "directives": Array [],
      "kind": "FragmentDefinition",
      "name": Object {
        "kind": "Name",
        "value": "MyFragment",
      },
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "id",
            },
            "selectionSet": undefined,
          },
          Object {
            "alias": undefined,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "name",
            },
            "selectionSet": undefined,
          },
        ],
      },
      "typeCondition": Object {
        "kind": "NamedType",
        "name": Object {
          "kind": "Name",
          "value": "User",
        },
      },
    },
    Object {
      "directives": Array [],
      "kind": "FragmentDefinition",
      "name": Object {
        "kind": "Name",
        "value": "TodoFragment",
      },
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "id",
            },
            "selectionSet": undefined,
          },
        ],
      },
      "typeCondition": Object {
        "kind": "NamedType",
        "name": Object {
          "kind": "Name",
          "value": "Todo",
        },
      },
    },
  ],
}
`;
